compiler_opt/distributed/local/buffered_scheduler.py

```diff
@@ -0,0 +1,82 @@
+# coding=utf-8
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""An optimal push-pull-based load balancer which attempts to maintain at least
+`buffer` tasks assigned to each assignee.
+"""
+
+import concurrent.futures
+import threading
+
+from typing import List, Callable, TypeVar
+
+
+class DeadlockDetected(Exception):
+
+  def __init__(self):
+    Exception.__init__(self)
+
+
+def raise_deadlock(*args, **kwargs):
+  # This is raised if attempting to call .result() on _NEVER_FUTURE
+  raise DeadlockDetected()
+
+
+_NEVER_FUTURE = concurrent.futures.Future()  # A Future that never completes.
+_NEVER_FUTURE.result = raise_deadlock
+T = TypeVar('T')
+
+
+def schedule(assignments: List[Callable[[T], concurrent.futures.Future]],
```

Comment:
you may want this to be Callable[[T], WorkerFuture] - see the comment before about Dask's futures